yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -228,6 +228,7 @@ private[spark] class ApplicationMaster(
   }
 
   private def sparkContextStopped(sc: SparkContext) = {
+    sc.requestTotalExecutors(0, 0, Map.empty)
```

Comment:
So, this doesn't really fix the issue, just makes it less likely. `sc.requestTotalExecutors` is asynchronous, so the allocator may still run before the message actually arrives in `AMEndpoint.receiveAndReply`.

That being said, I'm not sure the race is completely fixable. But you could get closer by putting the code currently in the `RequestExecutors` message handler in `AMEndpoint` in a separate method, and calling it from here.